crates/engine/tree/src/tree/precompile_cache.rs

```diff
@@ -0,0 +1,171 @@
+//! Contains a precompile cache that is backed by a moka cache.
+
+use reth_evm::precompiles::{DynPrecompile, Precompile};
+use revm::precompile::{PrecompileOutput, PrecompileResult};
+use revm_primitives::Bytes;
+use std::sync::Arc;
+
+/// Cache for precompiles, for each input stores the result.
+#[derive(Debug, Clone)]
+pub struct PrecompileCache(
+    Arc<mini_moka::sync::Cache<CacheKey, CacheEntry, alloy_primitives::map::DefaultHashBuilder>>,
+);
+
+impl Default for PrecompileCache {
+    fn default() -> Self {
+        Self(Arc::new(
+            mini_moka::sync::CacheBuilder::new(100_000)
+                .build_with_hasher(alloy_primitives::map::DefaultHashBuilder::default()),
+        ))
+    }
+}
+
+impl PrecompileCache {
+    fn get(&self, key: &CacheKey) -> Option<CacheEntry> {
+        self.0.get(key)
+    }
+
+    fn insert(&self, key: CacheKey, value: CacheEntry) {
+        self.0.insert(key, value);
+    }
+
+    fn weighted_size(&self) -> u64 {
+        self.0.weighted_size()
+    }
+}
+
+/// Cache key, precompile call input.
+#[derive(Debug, Clone, PartialEq, Eq, Hash)]
+pub struct CacheKey(alloy_primitives::Bytes);
+
+/// Cache entry, precompile successful output.
+#[derive(Debug, Clone, PartialEq, Eq)]
+pub struct CacheEntry(PrecompileOutput);
+
+impl CacheEntry {
+    const fn gas_used(&self) -> u64 {
+        self.0.gas_used
+    }
+
+    fn to_precompile_result(&self) -> PrecompileResult {
+        Ok(self.0.clone())
+    }
+}
+
+/// A cache for precompile inputs / outputs.
+#[derive(Debug)]
+pub(crate) struct CachedPrecompile {
+    /// Cache for precompile results and gas bounds.
+    cache: PrecompileCache,
+    /// The precompile.
+    precompile: DynPrecompile,
+    /// Cache metrics.
+    metrics: CachedPrecompileMetrics,
+}
+
+impl CachedPrecompile {
+    /// `CachedPrecompile` constructor.
+    pub(crate) fn new(precompile: DynPrecompile, cache: PrecompileCache) -> Self {
+        Self { precompile, cache, metrics: Default::default() }
+    }
+
+    pub(crate) fn wrap(precompile: DynPrecompile, cache: PrecompileCache) -> DynPrecompile {
+        let wrapped = Self::new(precompile, cache);
+        move |data: &[u8], gas_limit: u64| -> PrecompileResult { wrapped.call(data, gas_limit) }
+            .into()
+    }
+
+    fn increment_by_one_precompile_cache_hits(&self) {
+        self.metrics.precompile_cache_hits.increment(1);
+    }
+
+    fn increment_by_one_precompile_cache_misses(&self) {
+        self.metrics.precompile_cache_misses.increment(1);
+    }
+
+    fn increment_by_one_precompile_errors(&self) {
+        self.metrics.precompile_errors.increment(1);
+    }
+
+    fn update_precompile_cache_size(&self) {
+        let new_size = self.cache.weighted_size();
+        self.metrics.precompile_cache_size.set(new_size as f64);
+    }
+}
+
+impl Precompile for CachedPrecompile {
+    fn call(&self, data: &[u8], gas_limit: u64) -> PrecompileResult {
+        let key = CacheKey(Bytes::copy_from_slice(data));
```

Comment:
we can solve this more efficiently because we only need `Borrow` here, but we can take care of this seperately.